### PR TITLE
Re-implements as modular server using Express

### PR DESCRIPTION
**Pull request by DougReeder**
### Features/Fixes

- streaming transfer (documents don't need to fit in memory)
- JSON Web Tokens (JWT) so permissions can be validated in memory, without a round trip to a server or even reading a file
- S3-compatible storage (AWS S3 allows up to 5 TB)
- Bug Fix: correctly handles If-None-Match with ETag
- Bug Fix: returns empty listing for nonexistent folder
- Implements current spec: draft-dejong-remotestorage-22
- based on Express, for much cleaner code & standard extension interface
    - easier to reason about
    - stores are standard Express handlers
    - new kinds of authentication will be straighforward
    - Middleware exists for almost anything that is worth doing in an Express server

### Support for other extensions

Allows the extensibility of #88 and #89 and armadietto+lucchetto in a standard way.  Bespoke versions can be implemented by copying `appFactory.js` and adding new middleware. It will be much easier to say "yes" to extensions with a limited audience. Middleware for storage quotas, rate limiting and in-app purchase can be in the archive, but don't have to be in the default configuration. 

#### Store

To run the server or test S3-compatible storage, an S3-compatible instance must be running and environment variables set. See `notes/S3-streaming-store.md`.  To run automated tests:
```
node node_modules/mocha/bin/mocha.js spec/streaming_handler/S3Handler.spec.js
```

- AWS fully works
- Min.IO is usable, but has a leaky abstraction — the Min.IO web Console only displays the root folder & deleting a user doesn't fully complete.
- OpenIO fails the simultaneous deletion test — it shouldn't be used now, but perhaps some day it will be useable
- Could use testing with Google Cloud Storage, [DigitalOcean Spaces](https://www.digitalocean.com/products/spaces), etc.
- Could use more testing with files larger than 1GB — see the "transfers very large files" test

#### Automated Tests

The original automated tests are still in `spec/armadietto`. They validate that the monolithic server still functions the same.

The automated tests have been copied copied and reworked to allow testing the modular server against the same tests as the monolithic server. The tests starting with `a_` in `spec/armadietto` set up the monolithic server and call the tests in `spec`.  The tests starting with `m_` in `spec/modular` set up the modular server and call the tests in `spec`.  Thus, the tests validate that the modular server behaves the same as the monolithic server, aside from a few edge cases.

### CODE REVIEW

**Comment by raucao** [comment]
Just a note for self-hosting and S3: [Garage](https://garagehq.deuxfleurs.fr/) is also a great choice that is free software and not geared towards enterprise customers. We're currently beta-testing a new remoteStorage service using that over at [Kosmos](https://kosmos.org/).

File: lib/appFactory.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  if (basePath && !basePath.startsWith('/')) { basePath = '/' + basePath; }
+
+  const app = express();
+  app.locals.basePath = basePath;
+
+  // view engine setup
+  app.engine('.html', require('ejs').__express);
+  app.engine('.xml', require('ejs').__express);
+  app.set('view engine', 'html');
+  app.set('views', path.join(__dirname, 'views'));
```
Comment (AddictedToBattlestar) [inline lib/appFactory.js:23]: In my own work I haven't had to build an implementation of express for rendering needs.  I have used it only for API request/response handling where only data is involved.

From what I am parsing here.  If you are trying to tell express to render html and xml files render using the ejs rendering engine ... and for all other file types render using the html rendering engine, then I suppose this will work as you are hoping.

I would recommend just having the ejs rendering engine handle all content being passed here as I think that will just be simpler to implement and hopefully simpler to maintain (`app.set('view engine', 'ejs');`)

File: lib/appFactory.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        childSrc: ['\'none\''],
+        connectSrc: ['\'none\''],
+        baseUri: ['\'self\''],
+        frameAncestors: ['\'none\''],
+        formAction: (process.env.NODE_ENV === 'production' ? ['https:'] : ['https:', 'http:']), // allows redirect to any RS app
+        upgradeInsecureRequests: []
+      }
+    }
+  }));
+  app.use(express.urlencoded({ extended: false }));
```
Comment (AddictedToBattlestar) [inline lib/appFactory.js:50]: I am curious why you chose to not have URL encoding enabled for this app.  I normally am thankful that express automatically handles this concern.

File: lib/appFactory.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  app.use(`${basePath}/assets`, express.static(path.join(__dirname, 'assets')));
+
+  app.use(`${basePath}/`, indexRouter);
+
+  app.use(`${basePath}/signup`, signupRouter);
+
+  app.use([`${basePath}/.well-known`, `${basePath}/webfinger`], webFingerRouter);
+
+  app.use(`${basePath}/oauth`, oAuthRouter(hostIdentity, jwtSecret));
```
Comment (AddictedToBattlestar) [inline lib/appFactory.js:59]: It's interesting your choice of using the routing middleware implementation in express.  As a matter of preference, I like keeping all of my routing specifications in one file and then delegate the functional implementations to external "Controller" files for each endpoint being declared.  Only when the overall routing structure gets too cumbersome do I shunt sections to routing middleware.

I see no issues with how you implemented this.  I just wanted to provide a difference of opinion in my approach versus yours.

File: lib/appFactory.js [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+  app.use(`${basePath}/storage`, storageCommon(hostIdentity, jwtSecret));
+  app.use(`${basePath}/storage`, store);
+
+  // catches paths not handled and returns Not Found
+  app.use(basePath, function (req, res, next) {
+    const name = req.path.slice(1);
+    errorPage(req, res, 404, { title: 'Not Found', message: `“${name}” doesn't exist` });
+  });
```
Comment (AddictedToBattlestar) [inline lib/appFactory.js:67]: Nice.  It is still surprising to me how many apps I have worked on miss this simple fall through routing handler.

File: lib/routes/index.js [hunk 5]
```diff
@@ -0,0 +1,13 @@
+const express = require('express');
+const router = express.Router();
+const { getHost } = require('../util/getHost');
+
+/* GET home page. */
+router.get('/', function (req, res) {
+  res.render('index.html', {
+    title: 'Welcome',
+    host: getHost(req)
```
Comment (AddictedToBattlestar) [inline lib/routes/index.js:9]: I don't see a `title` nor a `host` parameter in the index.html template.

In looking through some of your other templates this is continuing to be the case.  I am assuming that they are special parameters for the rendering engine to set the page title and something in regards to the host?

File: lib/appFactory.js [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+  app.use(`${basePath}/`, indexRouter);
+
+  app.use(`${basePath}/signup`, signupRouter);
+
+  app.use([`${basePath}/.well-known`, `${basePath}/webfinger`], webFingerRouter);
+
+  app.use(`${basePath}/oauth`, oAuthRouter(hostIdentity, jwtSecret));
+  app.use(`${basePath}/storage`, storageCommon(hostIdentity, jwtSecret));
+  app.use(`${basePath}/storage`, store);
```
Comment (AddictedToBattlestar) [inline lib/appFactory.js:61]: It's odd from a naming perspective that you didn't name `storageCommon` and `store` with the same "...Router" suffix.  At least they are in this routes directory to callout how they are implemented here.

**Review by AddictedToBattlestar — COMMENTED** [review]
I have only really looked over the changes in `AppFactory` as well as some of the files that are imported there to support how express is setup.

I hope my feedback here is helpful regardless of this partial review for the PR.

File: lib/appFactory.js (hunk 1 above)
Comment (DougReeder) [reply lib/appFactory.js:23]: We actually are having ejs handle all rendering; `app.set('view engine', 'html');` says, when no extension is given, to assume the extension is `.html` (and thus to use EJS).
https://expressjs.com/en/api.html#app.set

File: lib/appFactory.js (hunk 3 above)
Comment (DougReeder) [reply lib/appFactory.js:59]: The main reason is sharing helper functions — for example, in `routes/oauth.js`, there's a helper function `validOAuthRequest` that is used by both routes:

```javascript
  router.get('/:username',
    redirectToSSL,
    formOrQueryData,
    sanityCheckUsername,
    validOAuthRequest,
    function (req, res) {
        // ...
    });

  router.post('/',
    secureRequest,
    formOrQueryData,
    sanityCheckUsername,
    validOAuthRequest,
    async function (req, res) {
        // ...
    }
  );

  function validOAuthRequest (req, res, next) {
        // ...
    next();
  }


```

File: lib/routes/index.js (hunk 5 above)
Comment (DougReeder) [reply lib/routes/index.js:9]: They are used by `header.html`, a template which is prepended to all of the page templates.

File: lib/appFactory.js (hunk 2 above)
Comment (DougReeder) [reply lib/appFactory.js]: That line actually does enable standard URL encoding.  I've enabled extended URL encoding (objects, etc), so we can use that where appropriate: https://github.com/remotestorage/armadietto/commit/081a8edea29a8d3ddbe00b58513e622a043dad27

File: lib/appFactory.js (hunk 6 above)
Comment (DougReeder) [reply lib/appFactory.js]: You're right that the naming is inconsistent; I have fixed that: https://github.com/remotestorage/armadietto/commit/081a8edea29a8d3ddbe00b58513e622a043dad27

**Comment by DougReeder** [comment]
> I have only really looked over the changes in `AppFactory` as well as some of the files that are imported there to support how express is setup.
> 
> I hope my feedback here is helpful regardless of this partial review for the PR.

Thanks for the feedback!  Having another set of eyes gives me confidence we're on the right course.

**Comment by DougReeder** [comment]
I would really like to have more feedback on a major change like this, but as this fixes several bugs and removes the obstacles to desired improvements (and doesn't foreclose development on the monolithic server), I will merge this in a week, unless there are objections.
